benchexec/container.py

```diff
@@ -723,28 +763,130 @@ def make_overlay_mount(mount, lower, upper, work):
         mount,
         b"overlay",
         0,
         b"lowerdir="
-        + escape(lower)
+        + escape_overlayfs_parameters(lower)
+        + b",upperdir="
+        + escape_overlayfs_parameters(upper)
+        + b",workdir="
+        + escape_overlayfs_parameters(work),
+    )
+
+
+def check_use_fuse_overlayfs(mount_base, dir_modes):
+    """
+    Check whether an overlay mountpoint requires the use of fuse-overlayfs
+    by determining if there are any sub-mounts below it.
+    """
+    mount_points = [
+        (full_mountpoint, fstype)
+        for _unused_source, full_mountpoint, fstype, _options in get_mount_points()
+        if util.path_is_below(full_mountpoint, mount_base)
+    ]
+
+    for full_mountpoint, fstype in mount_points:
+        mountpoint = full_mountpoint[len(mount_base) :] or b"/"
+        mode = determine_directory_mode(dir_modes, mountpoint, fstype)
+
+        if not mode or not os.path.exists(mountpoint):
+            continue
+
+        if mode == DIR_OVERLAY:
+            # Check if there are any sub-mounts within the current overlay mount point
+            for sub_mountpoint, _unused_fstype in mount_points:
+                if (
+                    util.path_is_below(sub_mountpoint, mountpoint)
+                    and sub_mountpoint != mountpoint
+                ):
+                    return True
+
+    return False
+
+
+@contextlib.contextmanager
+def permitted_cap_as_ambient():
+    try:
+        """
+        Transfer all permitted capabilities to the inheritable set
+        and raise them in the ambient set if effective.
+        """
+        header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
+        data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()
+
+        libc.capget(header, data)
+
+        original_inheritable = [data[0].inheritable, data[1].inheritable]
+        data[0].inheritable = data[0].permitted
+        data[1].inheritable = data[1].permitted
+
+        libc.capset(header, data)
+
+        effective = (data[1].effective << 32) | data[0].effective
+        cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
+        for cap in range(cap_last_cap + 1):
+            if effective & (1 << cap):
+                libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_RAISE, cap, 0, 0)
+        yield
+    finally:
+        """
+        Drop all ambient capabilities by removing them from the ambient set,
+        and undo changes made to inheritable set.
+        """
+        cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
+        for cap in range(cap_last_cap + 1):
+            libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_LOWER, cap, 0, 0)
+
+        header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
+        data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()
+
+        libc.capget(header, data)
+        data[0].inheritable, data[1].inheritable = original_inheritable
+        libc.capset(header, data)
```

Comment:
Now we can simplify this: common code like determining `cap_last_cap` can be moved up before the `try`. I think we do not even need to call `capget` twice, because we can also reuse header and data. So we prepare everything at the beginning of the function, then inside the `try` call `capset` and `prctl`, and in the `finally` part also only just `capset` after resetting `data` and `prctl`.